### PR TITLE
Unify response for requests with invalid ids

### PR DESCRIPTION
**Pull request by IKarbowiak**
- For queries:
  -  with valid `ID` but object not found - return `null`
  - with valid `ID` of different object type - return `null`
   - with invalid `ID` - raise GraphQLError
- For mutations:
  -  with valid `ID` but object not found - return `ValidationError`
  -  with invalid `ID` - raise `ValidationError`
  - with valid `ID` different object type - raise `ValidationError`

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-3511 Errors when resolving nodes by ID with invalid ID](https://app.clickup.com/t/2549495/SALEOR-3511)

**Review by korycins — APPROVED** [review]

File: saleor/graphql/core/mutations.py [hunk 1]
```diff
@@ -154,13 +154,27 @@ def get_node_by_pk(
...
 
+    @classmethod
+    def get_global_id_or_error(
+        cls, id: str, only_type: Union[ObjectType, str] = None, field: str = "id"
+    ):
+        try:
+            _object_type, pk = from_global_id_or_error(id, only_type, raise_error=True)
+        except GraphQLError as e:
+            raise ValidationError(
+                {field: ValidationError(str(e), code="graphql_error")}
```
[the remark this answers is not in the document]
Comment (IKarbowiak) [reply saleor/graphql/core/mutations.py:165]: I did it in the same way as it's done for other methods in `BaseMutation` like `get_global_id_or_error`. But maybe we should change the behavior for all of them and provide an error class enum as a parameter.

[the remark this answers is not in the document]
Comment (maarcingebala) [reply saleor/graphql/core/mutations.py:165]: I think we should have a hierarchy of those error classes, with a `BaseError` enum which defines all codes that are common, such a `graphql_error`. In generic places like this one, we could then use the base error.

File: saleor/graphql/csv/tests/mutations/test_export_products.py [hunk 2]
```diff
@@ -203,6 +204,53 @@ def test_export_products_mutation_ids_scope(
...
+):
+    query = EXPORT_PRODUCTS_MUTATION
+
+    products = product_list[:2]
+
+    ids = []
+    pks = set()
+    for product in products:
+        pks.add(str(product.pk))
+        ids.append(graphene.Node.to_global_id("ProductVariant", product.pk))
```
Comment (tomaszszymanski129) [inline saleor/graphql/csv/tests/mutations/test_export_products.py:223]:
```suggestion
    ids = [graphene.Node.to_global_id("ProductVariant", product.pk) for product in products]
```
I don't see `pks` set used anywhere?

File: saleor/graphql/csv/tests/mutations/test_export_products.py [hunk 3]
```diff
@@ -284,6 +332,180 @@ def test_export_products_mutation_with_warehouse_and_attribute_ids(
...
+    channel_PLN,
+    permission_manage_products,
+    permission_manage_apps,
+):
+    query = EXPORT_PRODUCTS_MUTATION
+
+    products = product_list[:2]
+
+    ids = []
+    pks = set()
```
Comment (tomaszszymanski129) [inline saleor/graphql/csv/tests/mutations/test_export_products.py:350]: same here, this is not used anywhere

File: saleor/graphql/giftcard/resolvers.py [hunk 4]
```diff
@@ -1,13 +1,8 @@
...
-
-from ...core.tracing import traced_resolver
 from ...giftcard import models
-from .types import GiftCard
 
 
-@traced_resolver
-def resolve_gift_card(info, gift_card_global_id):
-    return graphene.Node.get_node_from_global_id(info, gift_card_global_id, GiftCard)
+def resolve_gift_card(gift_card_id):
```
Comment (tomaszszymanski129) [inline saleor/graphql/giftcard/resolvers.py:4]: just a thought but do we have any convention for naming arguments of resolvers? In this PR i saw we sometimes use `id` and sometimes `thing_id`

File: saleor/graphql/warehouse/schema.py [hunk 5]
```diff
@@ -46,8 +47,8 @@ class WarehouseQueries(graphene.ObjectType):
     @traced_resolver
     def resolve_warehouse(self, info, **data):
         warehouse_pk = data.get("id")
-        warehouse = graphene.Node.get_node_from_global_id(info, warehouse_pk, Warehouse)
-        return warehouse
+        _, id = from_global_id_or_error(warehouse_pk, Warehouse)
+        return models.Warehouse.objects.filter(id=id).first()
```
Comment (tomaszszymanski129) [inline saleor/graphql/warehouse/schema.py:51]: Maybe we should extract ORM to dedicated resolvers file like we've done in many other places?

**Review by tomaszszymanski129 — APPROVED** [review]
Few things to consider, overall - nicely done!

**Review by maarcingebala — APPROVED** [review]
I'm impressed with the size of this PR - the changes look great ;) What bothers me is the number of different utility functions we have to resolve objects or IDs:
- `get_node_or_error`
- `get_global_id_or_error`
- `get_node_by_pk`
- `from_global_id_or_error` 

But if we were to change it, it's again a lot of work and requires a new PR. For this one, I was only thinking about making `get_node_by_pk` "private" (by renaming it to `_get_node_by_pk`)  as it is only used internally in `BaseMutation`.

File: saleor/graphql/giftcard/resolvers.py (hunk 4 above)
Comment (IKarbowiak) [reply saleor/graphql/giftcard/resolvers.py:4]: I think we don't have a naming convention but we can introduce it.

File: saleor/graphql/core/mutations.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (IKarbowiak) [reply saleor/graphql/core/mutations.py:165]: I created a task for that: https://app.clickup.com/t/mk6d2b.

**Review by tomaszszymanski129 — APPROVED** [review]

**Pull request merged**